bokehjs/src/lib/models/mappers/categorical_mapper.ts

```diff
@@ -1,48 +1,43 @@
 import type {Mapper} from "./mapper"
-import type {FactorSeq, Factor, L1Factor, L2Factor, L3Factor} from "../ranges/factor_range"
-import type {Arrayable, ArrayableOf} from "core/types"
+import type {FactorSeq, Factor} from "../ranges/factor_range"
+import type {Arrayable} from "core/types"
 import {index_of, find_index} from "core/util/arrayable"
-import {isString} from "core/util/types"
+import {isArray} from "core/util/types"
 import type * as p from "core/properties"
 
 export function _cat_equals(a: ArrayLike<unknown>, b: ArrayLike<unknown>): boolean {
-  if (a.length != b.length)
+  if (a.length != b.length) {
     return false
+  }
 
-  for (let i = 0, end = a.length; i < end; i++) {
+  const n = a.length
+  for (let i = 0; i < n; i++) {
```

Comment:
`a.length` is recomputed in each iteration, so it's an idiom to either write `for (let i, n = a.length, ...)` (for those who don't mind `let`) and otherwise to move the length computation outside the loop like here. Sadly any attempt at using zip-style iterator will degrade performance considerably.